pkg/distributor/query.go

```diff
@@ -173,3 +191,56 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 
 	return result, nil
 }
+
+// queryIngesterTSDBStream queries the ingesters using the TSDB streaming API
+func (d *Distributor) queryIngesterTSDBStream(ctx context.Context, replicationSet ring.ReplicationSet, req *client.QueryRequest) ([]client.TimeSeries, error) {
+	// Fetch samples from multiple ingesters
+	results, err := replicationSet.Do(ctx, d.cfg.ExtraQueryDelay, func(ing *ring.IngesterDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+		ingesterQueries.WithLabelValues(ing.Addr).Inc()
+
+		stream, err := client.(ingester_client.IngesterClient).QueryTSDBStream(ctx, req)
+		if err != nil {
+			ingesterQueryFailures.WithLabelValues(ing.Addr).Inc()
+			return nil, err
+		}
+		defer stream.CloseSend()
+
+		var result []*ingester_client.QueryTSDBStreamResponse
+		for {
+			series, err := stream.Recv()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, err
+			}
+			result = append(result, series)
+		}
+		return result, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	hashToSeries := map[model.Fingerprint]ingester_client.TimeSeries{}
+	for _, result := range results {
+		for _, response := range result.([]*ingester_client.QueryTSDBStreamResponse) {
+			for _, series := range response.Timeseries {
+				hash := client.FastFingerprint(series.Labels)
```

Comment:
This usage of fast fingerprint introduces a bug (it's also present in `queryIngesterStream`), where different timeseries with colliding fast fingerprint are merged together. We should fix both places.

Reply:
What's the correct function to user here?

Reply:
Easiest solution is to skip any fingerprinting, and just convert labels to a string, and use that string as a key to hash map.

Reply:
Note: if you are going to make this change, it should be done in `queryIngesterStream` as well. Perhaps a separate PR is a better idea.

Reply:
I think this should be brought up in a separate PR. Going to defer for now.